File: src/DBuilder.Geometry/ProjectedFrustum2D.cs
// ABOUTME: 2D-projected view frustum ported from UDB Source/Core/Geometry/ProjectedFrustum2D.cs.
// ABOUTME: Builds a planar frustum polygon plus bounding circle for fast visibility culling.

/*
 * Copyright (c) 2007 Pascal vd Heiden, www.codeimp.com
 * This program is released under GNU General Public License
 */

using System;

namespace DBuilder.Geometry;

public class ProjectedFrustum2D
{
    private readonly float near;
    private readonly float far;
    private readonly float fov;
    private readonly Vector2D pos;
    private readonly float xyangle;
    private readonly float zangle;

    private readonly Line2D[] lines;

    private readonly Vector2D center;
    private readonly float radius;

    public float Near => near;
    public float Far => far;
    public float Fov => fov;
    public Vector2D Position => pos;
    public float XYAngle => xyangle;
    public float ZAngle => zangle;
    public Line2D[] Lines => lines;
    public Vector2D Center => center;
    public float Radius => radius;

    public ProjectedFrustum2D(Vector2D pos, float xyangle, float zangle, float near, float far, float fov)
    {
        Vector2D[] forwards = new Vector2D[4];
        Vector2D[] downwards = new Vector2D[4];
        Vector2D[] corners = new Vector2D[4];

        this.pos = pos;
        this.xyangle = xyangle;
        this.zangle = zangle;
        this.near = near;
        this.far = far;
        this.fov = fov;

        // Make the corners for a forward frustum
        // Order: Left-Far, Right-Far, Left-Near, Right-Near
        float fovhalf = fov * 0.5f;
        float fovhalfcos = (float)Math.Cos(fovhalf);
        float farsidelength = far / fovhalfcos;
        float nearsidelength = near / fovhalfcos;
        forwards[0] = pos + Vector2D.FromAngle(xyangle - fovhalf, farsidelength);
        forwards[1] = pos + Vector2D.FromAngle(xyangle + fovhalf, farsidelength);
        forwards[2] = pos + Vector2D.FromAngle(xyangle - fovhalf, nearsidelength);
        forwards[3] = pos + Vector2D.FromAngle(xyangle + fovhalf, nearsidelength);

        // Corners for a downward frustum
        float farradius = (float)(far * 0.5f * Angle2D.SQRT2);
        downwards[0] = pos + Vector2D.FromAngle(xyangle - Angle2D.PI * 0.25f, farradius);
        downwards[1] = pos + Vector2D.FromAngle(xyangle + Angle2D.PI * 0.25f, farradius);
        downwards[2] = pos + Vector2D.FromAngle(xyangle - Angle2D.PI * 0.75f, farradius);
        downwards[3] = pos + Vector2D.FromAngle(xyangle + Angle2D.PI * 0.75f, farradius);

        // Interpolate between forward and downward based on z angle
        float d = Math.Abs((float)Math.Sin(zangle));
        corners[0] = forwards[0] * (1.0f - d) + downwards[0] * d;
        corners[1] = forwards[1] * (1.0f - d) + downwards[1] * d;
        corners[2] = forwards[2] * (1.0f - d) + downwards[2] * d;
        corners[3] = forwards[3] * (1.0f - d) + downwards[3] * d;

        // Frustum lines (all oriented so that their right side is inside the frustum)
        lines = new Line2D[4];
        lines[0] = new Line2D(corners[2], corners[0]);
        lines[1] = new Line2D(corners[1], corners[3]);
        lines[2] = new Line2D(corners[3], corners[2]);
        lines[3] = new Line2D(corners[0], corners[1]);

        center = (corners[0] + corners[1] + corners[2] + corners[3]) * 0.25f;

        float radius2 = 0.0f;
        for (int i = 0; i < corners.Length; i++)
        {
            float distance2 = (float)Vector2D.DistanceSq(center, corners[i]);
            if (distance2 > radius2) radius2 = distance2;
        }
        radius = (float)Math.Sqrt(radius2);
    }

    // Checks if a specified circle is intersecting the frustum
    // NOTE: This checks only against the actual frustum and does not use the frustum circle.
    public bool IntersectCircle(Vector2D circlecenter, float circleradius)
    {
        for (int i = 0; i < lines.Length; i++)
        {
            if (lines[i].GetSideOfLine(circlecenter) < 0)
            {
                // Center is outside the frustum; check overlap.
                if (lines[i].GetDistanceToLineSq(circlecenter, false) > (circleradius * circleradius)) return false;
            }
        }

        return true;
    }

    public bool IntersectBox(Vector2D boxcenter, double halfwidth, double halfheight)
    {
        for (int i = 0; i < lines.Length; i++)
        {
            Line2D line = lines[i];
            double dx = line.v2.x - line.v1.x;
            double dy = line.v2.y - line.v1.y;
            double a = -dy;
            double b = dx;
            double d = -(line.v1.x * a + line.v1.y * b);
            double e = halfwidth * Math.Abs(a) + halfheight * Math.Abs(b);
            double s = boxcenter.x * a + boxcenter.y * b + d;
            if (s + e < 0.0) return false;
        }

        return true;
    }
}

File: rust/crates/dbuilder-geometry/src/projected_frustum2d.rs
// ABOUTME: 2D-projected view frustum ported from DBuilder ProjectedFrustum2D.cs.
// ABOUTME: Preserves UDB's mixed float/double math; f32 stays f32 where C# used float.

/*
 * Copyright (c) 2007 Pascal vd Heiden, www.codeimp.com
 * This program is released under GNU General Public License
 */

use crate::angle2d;
use crate::line2d::Line2D;
use crate::vector2d::Vector2D;

pub struct ProjectedFrustum2D {
    near: f32,
    far: f32,
    fov: f32,
    pos: Vector2D,
    xyangle: f32,
    zangle: f32,

    lines: [Line2D; 4],

    center: Vector2D,
    radius: f32,
}

impl ProjectedFrustum2D {
    pub fn new(
        pos: Vector2D,
        xyangle: f32,
        zangle: f32,
        near: f32,
        far: f32,
        fov: f32,
    ) -> ProjectedFrustum2D {
        // Make the corners for a forward frustum
        // Order: Left-Far, Right-Far, Left-Near, Right-Near
        let fovhalf: f32 = fov * 0.5;
        let fovhalfcos: f32 = (fovhalf as f64).cos() as f32;
        let farsidelength: f32 = far / fovhalfcos;
        let nearsidelength: f32 = near / fovhalfcos;
        let forwards = [
            pos + Vector2D::from_angle_with_length(
                (xyangle - fovhalf) as f64,
                farsidelength as f64,
            ),
            pos + Vector2D::from_angle_with_length(
                (xyangle + fovhalf) as f64,
                farsidelength as f64,
            ),
            pos + Vector2D::from_angle_with_length(
                (xyangle - fovhalf) as f64,
                nearsidelength as f64,
            ),
            pos + Vector2D::from_angle_with_length(
                (xyangle + fovhalf) as f64,
                nearsidelength as f64,
            ),
        ];

        // Corners for a downward frustum
        // C#: (float)(far * 0.5f * Angle2D.SQRT2) -- f32 product widened by the double constant.
        let farradius: f32 = ((far * 0.5) as f64 * angle2d::SQRT2) as f32;
        let downwards = [
            pos + Vector2D::from_angle_with_length(
                xyangle as f64 - angle2d::PI * 0.25f32 as f64,
                farradius as f64,
            ),
            pos + Vector2D::from_angle_with_length(
                xyangle as f64 + angle2d::PI * 0.25f32 as f64,
                farradius as f64,
            ),
            pos + Vector2D::from_angle_with_length(
                xyangle as f64 - angle2d::PI * 0.75f32 as f64,
                farradius as f64,
            ),
            pos + Vector2D::from_angle_with_length(
                xyangle as f64 + angle2d::PI * 0.75f32 as f64,
                farradius as f64,
            ),
        ];

        // Interpolate between forward and downward based on z angle
        let d: f32 = ((zangle as f64).sin() as f32).abs();
        let corners = [
            forwards[0] * (1.0f32 - d) as f64 + downwards[0] * d as f64,
            forwards[1] * (1.0f32 - d) as f64 + downwards[1] * d as f64,
            forwards[2] * (1.0f32 - d) as f64 + downwards[2] * d as f64,
            forwards[3] * (1.0f32 - d) as f64 + downwards[3] * d as f64,
        ];

        // Frustum lines (all oriented so that their right side is inside the frustum)
        let lines = [
            Line2D::new(corners[2], corners[0]),
            Line2D::new(corners[1], corners[3]),
            Line2D::new(corners[3], corners[2]),
            Line2D::new(corners[0], corners[1]),
        ];

        let center = (corners[0] + corners[1] + corners[2] + corners[3]) * 0.25;

        let mut radius2: f32 = 0.0;
        for corner in corners {
            let distance2 = Vector2D::distance_sq(center, corner) as f32;
            if distance2 > radius2 {
                radius2 = distance2;
            }
        }
        let radius = (radius2 as f64).sqrt() as f32;

        ProjectedFrustum2D {
            near,
            far,
            fov,
            pos,
            xyangle,
            zangle,
            lines,
            center,
            radius,
        }
    }

    pub fn near(&self) -> f32 {
        self.near
    }

    pub fn far(&self) -> f32 {
        self.far
    }

    pub fn fov(&self) -> f32 {
        self.fov
    }

    pub fn position(&self) -> Vector2D {
        self.pos
    }

    pub fn xy_angle(&self) -> f32 {
        self.xyangle
    }

    pub fn z_angle(&self) -> f32 {
        self.zangle
    }

    pub fn lines(&self) -> &[Line2D; 4] {
        &self.lines
    }

    pub fn center(&self) -> Vector2D {
        self.center
    }

    pub fn radius(&self) -> f32 {
        self.radius
    }

    // Checks if a specified circle is intersecting the frustum
    // NOTE: This checks only against the actual frustum and does not use the frustum circle.
    pub fn intersect_circle(&self, circlecenter: Vector2D, circleradius: f32) -> bool {
        for line in &self.lines {
            if line.get_side_of_line(circlecenter) < 0.0 {
                // Center is outside the frustum; check overlap.
                if line.get_distance_to_line_sq(circlecenter, false)
                    > (circleradius * circleradius) as f64
                {
                    return false;
                }
            }
        }

        true
    }

    pub fn intersect_box(&self, boxcenter: Vector2D, halfwidth: f64, halfheight: f64) -> bool {
        for line in &self.lines {
            let dx = line.v2.x - line.v1.x;
            let dy = line.v2.y - line.v1.y;
            let a = -dy;
            let b = dx;
            let d = -(line.v1.x * a + line.v1.y * b);
            let e = halfwidth * a.abs() + halfheight * b.abs();
            let s = boxcenter.x * a + boxcenter.y * b + d;
            if s + e < 0.0 {
                return false;
            }
        }

        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn frustum_builds_four_lines_and_positive_radius() {
        let f = ProjectedFrustum2D::new(
            Vector2D::new(0.0, 0.0),
            0.0,
            0.0,
            10.0,
            100.0,
            (std::f64::consts::PI / 2.0) as f32,
        );
        assert_eq!(4, f.lines().len());
        assert!(f.radius() > 0.0);
    }

    #[test]
    fn circle_inside_frustum_intersects() {
        // UDB's FromAngle(a) returns (sin a, -cos a), so xyangle=0 points along -Y.
        // A circle deep inside the frustum body is at (0, -50).
        let f = ProjectedFrustum2D::new(
            Vector2D::new(0.0, 0.0),
            0.0,
            0.0,
            10.0,
            100.0,
            (std::f64::consts::PI / 2.0) as f32,
        );
        assert!(f.intersect_circle(Vector2D::new(0.0, -50.0), 1.0));
    }

    #[test]
    fn circle_behind_camera_does_not_intersect() {
        // xyangle=0 points along -Y (sin 0, -cos 0); behind the camera is +Y.
        let f = ProjectedFrustum2D::new(
            Vector2D::new(0.0, 0.0),
            0.0,
            0.0,
            10.0,
            100.0,
            (std::f64::consts::PI / 4.0) as f32,
        );
        assert!(!f.intersect_circle(Vector2D::new(0.0, 10000.0), 1.0));
    }

    #[test]
    fn box_inside_frustum_intersects() {
        let f = ProjectedFrustum2D::new(
            Vector2D::new(0.0, 0.0),
            0.0,
            0.0,
            10.0,
            100.0,
            (std::f64::consts::PI / 2.0) as f32,
        );

        assert!(f.intersect_box(Vector2D::new(0.0, -50.0), 8.0, 8.0));
    }

    #[test]
    fn box_behind_camera_does_not_intersect() {
        let f = ProjectedFrustum2D::new(
            Vector2D::new(0.0, 0.0),
            0.0,
            0.0,
            10.0,
            100.0,
            (std::f64::consts::PI / 4.0) as f32,
        );

        assert!(!f.intersect_box(Vector2D::new(0.0, 1000.0), 8.0, 8.0));
    }
}
